chore: improve providence docs (#2408)
ing-bank/lion: docs/fundamentals/node-tools/providence-analytics/overview.md

- etc...

> Providence uses AST analysis for getting the most advanced analysis possible.

Author: ```suggestion
Providence uses abstract syntax trees (ASTs) to have the most advanced analysis possible.
```

It does this via the [oxc parser](https://oxc.rs/docs/guide/usage/parser.html), the quickest parser available today!

## Run

Providence expects an analyzer name that tells it what type of analysis to run:

```bash
npx providence analyze <analyzer-name>
```

By default Providence ships these analyzers:

- find-imports
- find-exports
- find-classes
- match-imports
- match-subclasses

> Let's say we run `find-exports`:

Author: ```suggestion
Let's say we run `find-imports`:
```

```bash
npx providence analyze find-imports
```

> Now it retrieves all relevant data about es module exports.

Author: ```suggestion
Now it retrieves all relevant data about es module imports.
```

Providence uses the concept of projects. A project is a piece of software to analyze:
usually an npm artifact or a git (mono-)repository. What all projects have in common,
is a package.json. From it, the following project data is derived:

- the name
- the version
- the files it uses for scanning. One of the following strategies is usually followed:
  - exportmap entrypoints (by 'expanding' package.json "exports" on file system)
  - npm files (it reads package.json "files" | .npmignore)
  - the git files (it reads .gitignore)
  - a custom defined list

For a "find" analyzer, there is one project involved (the target project).

We can specify it like this (we override the default current working directory):

```bash
npx providence analyze find-imports -t /importing/project
```

For a "match" analyzer, there is also a reference project.
Here we match the exports of the reference project (-r) against the imports of the target project (-t).

```bash
npx providence analyze match-imports -t /importing/project -r /exporting/project
```

## Utils

Providence comes with many tools for deep traversal of identifiers,
the (babel like) traversal of ast trees in oxc and swc and more.

> Also more generic utils for caching and performant globing come delivered with Providence.

Author: ```suggestion
Also more generic utils for caching and performant globbing come delivered with Providence.
```
